### PR TITLE
[core] Add specific tensor descriptor for Result

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Introduce shared tensor descriptor:
   - It has specific tensor names.
   - When specific tensor names set they will be used as tensor names otherwise the shred tensor names are used.
   - Specific tensor names will be added to shared descriptor (input descriptor), and transferred to new tensor descriptor if changed.
   
 - Use shared tensor descriptor for Result, which got specific tensor names.
 - Add dev API Extension interface for descriptor::Tensor for internal OV use.
 - Update `ReverseShapeAndTypeInfer` to use descriptor interface instead change properties directly. Update `PartialShape` of descriptor can change it in undefined state with `Shape` property.
 - Improves tensor names handling for model outputs introduced in #23713
 - Fix tensor names issue in `unroll_if` transformation (ticket: 114763)

### Tickets:
 - CVS-143105
 - CVS-146649
 - CVS-114763

### Blocked  by:
- intel-innersource/frameworks.ai.openvino.tests#1684 - fixes e2e tests

### CODE REVIEW

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 1]
```diff
@@ -940,14 +940,16 @@ def __call__(self, input):
...
-        assert ov_model.outputs[0].names == {"name1"}
+        assert "name1" in ov_model.outputs[0].names
 
         ov_model.validate_nodes_and_infer_types()
-        assert ov_model.outputs[0].names == {"name1"}
+        assert "name1" in ov_model.outputs[0].names
 
         ov_model.outputs[0].get_tensor().set_names({"name2"})
-        assert ov_model.outputs[0].names == {"name2"}
+        assert "name2" in ov_model.outputs[0].names
```
Comment (rkazants) [inline tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py:952]: why now tensor names is changed? We should have the single tensor name instead of several

**Review by rkazants — DISMISSED** [review]
why several tensor names are not clear

File: src/core/src/descriptor/tensor.cpp [hunk 2]
```diff
@@ -8,150 +8,294 @@
...
+
+private:
+    element::Type m_element_type;
+    ShapeInfo m_shape_info;
+    std::unordered_set<std::string> m_names;
+    std::unordered_set<std::string>::const_iterator m_name_it;
+    RTMap m_rt_map;
+    std::string m_legacy_name;
+
+    static decltype(m_name_it) find_new_any_name(const decltype(m_names)& names, decltype(m_name_it) name) {
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/descriptor/tensor.cpp]: Removed.

File: src/core/include/openvino/core/descriptor/tensor.hpp [hunk 3]
```diff
@@ -22,103 +22,107 @@ namespace ov {
...
 public:
+    /// \brief Creates Tensor descriptor
+    /// \param element_type Element type
+    /// \param pshape       Partial shape of tensor
+    /// \param names        Tensor names (optional default empty).
     Tensor(const element::Type& element_type,
            const PartialShape& pshape,
            const std::unordered_set<std::string>& names = {});
+
+    OPENVINO_DEPRECATED("This constructor is deprecated. Will be removed in 2025.0")
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/include/openvino/core/descriptor/tensor.hpp]: Changed to 2026.0

File: src/core/tests/type_prop/result.cpp [hunk 4]
```diff
@@ -51,3 +55,101 @@ TEST(type_prop, result_layout_invalid) {
...
+TEST_F(TypePropResultV0Test, set_specific_output_name_by_output) {
+    auto a = std::make_shared<Parameter>(element::f32, PartialShape::dynamic());
+    a->get_output_tensor(0).set_names({"input"});
+
+    auto result = make_op(a);
+
+    EXPECT_THAT(result->output(0).get_names(), UnorderedElementsAre("input"));
+    EXPECT_THAT(result->get_output_tensor(0).get_names(), UnorderedElementsAre("input"));
+
+    EXPECT_THAT(result->output(0).get_names(), UnorderedElementsAre("input"));
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/tests/type_prop/result.cpp]: No, removed.

**Review by t-jankowski — APPROVED** [review]

File: src/core/include/openvino/op/result.hpp [hunk 5]
```diff
@@ -13,6 +13,13 @@ namespace v0 {
 /// \brief Result operation.
 ///
 /// \ingroup ov_ops_cpp_api
+///
+/// The Result operator output is special output which share tensor with node connected to this node.
+///
+/// The Result's output names are visible as model outputs names.
+/// When set/add Result's output tensor names they will treat as this result specific names which will be appended to
+/// connect input tensor or transferred to new tensor connect Result to new input.
+/// When Result has not specific names the names from connected input will be used.
```
Comment (slyalin) [inline src/core/include/openvino/op/result.hpp:22]: It is hard to follow. Could you add more details and itemize all the variants?

File: tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py [hunk 6]
```diff
@@ -8,21 +8,46 @@
...
-    for t1 in tensors1:
-        t2_candidates = [t for t in tensors2 if t1.names & t.names]
-        assert len(t2_candidates) == 1
-        t2 = t2_candidates[0]
-        assert t1.names == t2.names
-        assert t1.get_partial_shape() == t2.get_partial_shape()
-        assert t1.get_element_type() == t2.get_element_type()
+def check_desc_tensors(expected_tensors, tensors):
+    # checks if tensors descriptors are same as expected
+    pattern_input = re.compile(R"input_restored.((past_key_values|present)\.(\d+)\.(key|value))")
```
Comment (jane-intel) [inline tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py:18]: Could you elaborate on why this change is needed? I'm a bit confused, are we backward compatible?

File: tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py [hunk 7]
```diff
@@ -8,21 +8,46 @@
...
+
+def check_result_desc_tensors(expected_tensors, tensors):
+    # checks if Result tensors descriptors are same as expected
+    pattern_restore_output = re.compile(R"output_restored.((past_key_values|present)\.(\d+)\.(key|value))")
+    pattern_output = re.compile(R"(present\.(\d+)\.(key|value))")
+
+    assert len(expected_tensors) == len(tensors)
+    for expected in expected_tensors:
+        # The `patch_stateful` in optimum use any name instead found key/value names OV will use names restore path
+        # Restore expected names to find tensor for compare (can be removed when HG optimum updated)
```
Comment (jane-intel) [inline tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py:41]: What is the problem with HG optimum? Does this mean that get_any_name returns different names now?

File: src/core/tests/type_prop/result.cpp [hunk 8]
```diff
@@ -51,3 +55,100 @@ TEST(type_prop, result_layout_invalid) {
...
+}
+
+TEST_F(TypePropResultV0Test, preserve_specific_name_on_input_replace) {
+    const auto a = std::make_shared<Parameter>(element::f32, PartialShape::dynamic());
+    a->get_output_tensor(0).set_names({"input a"});
+
+    const auto result = make_op(a);
+    result->output(0).set_names({"out"});
+
+    EXPECT_THAT(result->output(0).get_names(), UnorderedElementsAre("out"));
```
Comment (jane-intel) [inline src/core/tests/type_prop/result.cpp:141]: What is expected of result->input(0) here? Could you explain / add checks here?

File: src/core/src/preprocess/preprocess_impls.cpp [hunk 9]
```diff
@@ -325,11 +326,9 @@ void InputInfo::InputInfoImpl::dump(std::ostream& str,
 
 //----------- OutputInfoImpl ----------
 void OutputInfo::OutputInfoImpl::build(ov::ResultVector& results) {
-    std::shared_ptr<opset8::Result> result;
     auto node = m_output_node;
-    const auto start_out_node_names = node.get_tensor().get_names();
-    node.get_tensor().set_names({});
-    result = std::dynamic_pointer_cast<opset8::Result>(node.get_node_shared_ptr());
+    const auto result = std::dynamic_pointer_cast<opset8::Result>(node.get_node_shared_ptr());
```
Comment (jane-intel) [inline src/core/src/preprocess/preprocess_impls.cpp:330]: nit ov::as_type_ptr?

File: src/core/src/node.cpp [hunk 10]
```diff
@@ -472,8 +467,8 @@ ov::descriptor::Tensor& ov::Node::get_output_tensor(size_t i) const {
 
 ov::descriptor::Tensor& ov::Node::get_input_tensor(size_t i) const {
     OPENVINO_ASSERT(i < m_inputs.size(), idx_txt, i, out_of_range_txt);
-    descriptor::Input input = m_inputs[i];
-    return input.get_tensor();
+    auto& input = m_inputs[i];
+    return input.get_output().get_tensor();
```
Comment (jane-intel) [inline src/core/src/node.cpp:471]: Why this had to be changed? `input.get_tensor()` wouldn't work?

**Review by e-nugmanova — COMMENTED** [review]
Hello, thank you for this PR.
Could you share why CVS-115911 results weren't taken into account? It seems like the presentation and decision was made to have different approach.

Also, please clarify what happens with get_any_name -- why it starts to be incompatible?

File: src/core/src/node.cpp (hunk 10 above)
Comment (praasz) [reply src/core/src/node.cpp:471]: The first line it was just to remove unnecessary copy which was used to make a little hack (similar to `const_cast`) to remove consents from `input` as this function is `const`.
Without copy the connected node output tensor has to be returned which is not constant and it give better description what is returned . 

This is maybe separate topic but maybe `get_input_tensor` tensor should return const descriptor to avoid modification of descriptors which not belongs to this Node. The Node's inputs are not tensors owners.

File: src/core/tests/type_prop/result.cpp (hunk 8 above)
Comment (praasz) [reply src/core/tests/type_prop/result.cpp:141]:
```cpp
EXPECT_THAT(result->input(0).get_tensor().get_names(), UnorderedElementsAre("out", "input a"));
```
Check will be added

File: src/core/src/preprocess/preprocess_impls.cpp [hunk 11]
```diff
@@ -369,56 +368,36 @@ void OutputInfo::OutputInfoImpl::build(ov::ResultVector& results) {
...
     }
+
     if (orig_parent->get_output_size() == 1) {
         node.get_node_shared_ptr()->set_friendly_name(orig_parent->get_friendly_name());
-        reset_orig_friendly_name = true;
+
+        // Reset friendly name of input node to avoid names collision
+        // when there is at a new node inserted by post-processing steps
+        // If no new nodes are inserted by post-processing, then we need to preserve friendly name of input
+        // as it's required for old API correct work
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/preprocess/preprocess_impls.cpp:388]: For PPP compatibility mode it has to be kept (I don't want to change this functionality in this PR).

For IRv11 it can be considered as  a bug which is solved in [PR #23713](https://github.com/openvinotoolkit/openvino/pull/23713/files#diff-7c4d81a6d4bb639ff9af23a3a8f09cda9ec474f9fa1d6acf988c21b9c5ede74dR373). 

The Node's (layer) name should not be moved which was reported in [CVS-127482](https://jira.devtools.intel.com/browse/CVS-127482)

File: tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py (hunk 7 above)
Comment (praasz) [reply tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py:41]: The `get_any_name` is not reliable to use for testing to compare names when tensor can have many names or when some names are added to tensor.

As Result now not set its input names (they are not lost) the additional names can appear on compiled/transformed model. If there is many names get any name can give different name than on origin model and test fail but origin name is still on tensor.

This modification will only take stateless names on inputs and outputs and check if they are still present on model after transformation.

The HG issue was that it use any name and as there could be more names on tensor it pick the stateless pattern name or other. In case other name has been used the restore name path in transformation is used. This change will give more robust check whatever path is selected in transformation it will extract the names and prepare expected  and check if exists on input/outputs  on model after transformation.

The HG issue with any name has been fixed but OV can still use older version.

File: tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py (hunk 6 above)
Comment (praasz) [reply tests/model_hub_tests/transformation_tests/test_stateful_to_stateless_transformation.py:18]: Same as [here](https://github.com/openvinotoolkit/openvino/pull/25954#discussion_r1800909046)

**Comment by praasz** [comment]
> Hello, thank you for this PR. Could you share why [CVS-115911](https://jira.devtools.intel.com/browse/CVS-115911) results weren't taken into account? It seems like the presentation and decision was made to have different approach.
> 
> Also, please clarify what happens with get_any_name -- why it starts to be incompatible?
I've seen the CVS-115911 after I've done this PR, so I was not aware about it.

But I've done similar tries the best options was to apply changes in descriptor or make additional operator.
With Identity operator I've got some doubts (decide to try improve descriptor):

- It is required to add new operator
- There is add ne transformations to add/remove automatically between Result and other node
- When Identity added there has to be decision about current names which are Result's and which Result's input
- When Identity is removed the information which name was set to which Node is lost
- When Identity is connected to Result it could be required some dedicated handling from other nodes as Result is special

The current implementation is not making full names separation as Result's input has all names which allow to be backward compatible for (de)serialization.
To make full separation in descriptor will require addition effort and can be problematic for serialization.

Have this feature in descriptor then is not required to use additional operator (Identity can be useful for other purposes)
and there is no need to addition pre/post process model to add/remove them.

The functionality of `get_any_name` is not changed. It is not good to use in way that for some output on origin model we get any name then after model processing get again any name from same output then it maybe but must be same.
It depends how model is transformed as some outputs can get more names. Is better to `get_any_name` and check if this name is still in output after model processing.

**Review by alexsu52 — APPROVED** [review]
I confirm that this PR fixes NNCF layer-wise algorithms such as GPTQ.

**Review by e-nugmanova — APPROVED** [review]
